desub-current/src/decoder/decode_storage.rs

```diff
@@ -0,0 +1,330 @@
+	#[error("Expecting the same number of keys and hashers, but got {num_keys} keys and {num_hashers} hashers")]
+	KeysAndHashersDontLineUp { num_keys: usize, num_hashers: usize },
+	#[error("Type with id {0} expected in the metadata but not found")]
+	TypeNotFound(u32),
+	#[error("Couldn't decode the value associated with the hasher {0:?}: {1}")]
+	CouldNotDecodeHasherValue(frame_metadata::v14::StorageHasher, super::DecodeValueError),
+	#[error("Couldn't find a storage entry corresponding to the prefix hash provided in the data")]
+	PrefixNotFound,
+	#[error("Couldn't find a storage entry corresponding to the name hash provided in the data")]
+	NameNotFound,
+}
+
+impl StorageDecoder {
+	/// Call [`super::decode_storage()`] to construct a [`StorageDecoder`].
+	pub(super) fn generate_from_metadata(metadata: &Metadata) -> StorageDecoder {
+		let entries_by_hashed_prefix = metadata
+			.storage_entries()
+			.enumerate()
+			.map(|(index, entries)| {
+				let prefix_hash = twox_128(entries.prefix().as_bytes());
+				let entry_by_hashed_name = entries
+					.entries()
+					.enumerate()
+					.map(|(entry_index, entry)| {
+						let name_hash = twox_128(entry.name.as_bytes());
+						(name_hash, entry_index)
+					})
+					.collect();
+				(prefix_hash, StorageEntries { index, entry_by_hashed_name })
+			})
+			.collect();
+
+		StorageDecoder { entries_by_hashed_prefix }
+	}
+
+	/// Decode the SCALE encoded bytes representing a storage entry lookup. These conceptually take the
+	/// form `twox_128(prefix) + twox_128(name) + rest`, where `rest` depends on the storage entry we're
+	/// keying into, and may be nothing at all for plain storage locations, or hashed keys to access maps.
+	pub fn decode_key<'m, 'b>(
+		&self,
+		metadata: &'m Metadata,
+		bytes: &mut &'b [u8],
+	) -> Result<StorageEntry<'m, 'b>, StorageDecodeError> {
+		// Step 1: reverse-lookup the hashed prefix+name part of the key, and get
+		// details about this storage location from our metadata.
+		let location = self.decode_prefix_and_name_to_location(bytes)?;
+		let storage_entry = metadata.storage_entry(location);
+
+		let prefix_str = storage_entry.prefix;
+		let name_str = &*storage_entry.metadata.name;
+
+		// Step 2: use the details held in metadata to infer what form the rest of
+		// the bytes should take, and decode accordingly.
+		match &storage_entry.metadata.ty {
+			FrameStorageEntryType::Plain(ty) => {
+				// No more work to do here; our storage entry is a plain prefix+name entry,
+				// so return the details of it:
+				Ok(StorageEntry {
+					prefix: prefix_str.into(),
+					name: name_str.into(),
+					ty: *ty,
+					details: StorageEntryType::Plain,
+				})
+			}
+			FrameStorageEntryType::Map { hashers, key, value } => {
+				// We'll consume some more data based on the hashers.
+				// First, get the type information that we need ready.
+				let key =
+					metadata.types().resolve(key.id()).ok_or_else(|| StorageDecodeError::TypeNotFound(key.id()))?;
```

Comment:
Mmm, this error and the one below about keys and hashers not lining up could reasonably be converted to panics I think, since they suggest an internal issue with the metadata, and not something that the user could do anything about!

Reply:
I've converted the "internal" errors into panics; I _think_ that this is sensible, and better splits obvious user errors and our-metadata-is-forked errors.

(the assumption is that the metadata that they provide must itself be correct)